tests/external_executor/executors/logging_app.py

```diff
@@ -98,9 +103,16 @@ def run_loop(self, terminate_event):
                     self.do_get(stub, table, request, response)
                 else:
                     LOG.error(f"Unhandled request: {request.method} {request.uri}")
-                    stub.EndTx(response)
-                    continue
 
                 stub.EndTx(response)
 
         LOG.info("Ended executor loop")
+
+    def terminate(self):
+        target_uri = self.ccf_node.get_public_rpc_address()
+        with grpc.secure_channel(
+            target=target_uri,
+            credentials=self.credentials,
+        ) as channel:
+            stub = Service.KVStub(channel)
+            stub.Deactivate(Empty())
```

Comment:
What if this isn't called by the executor? Would the CCF node think that the executor is still alive and so try to dispatch work to it? Should there be a watchdog mechanism so that the executor is automatically deactivated if it hasn't responded for a while (perhaps we could make use of HTTP/2 ping frames for that)?

Reply:
That's what the close callback does - if the `Activate` stream is closed by the client (without calling `Deactivate` first), then the executor is deregistered on the server side by the close callback, and won't be dispatched any more work.

Reply:
Ah, I'd missed that, thanks. Do we have any plans to add more fine-grained health checks/stats for the external executor, e.g. to cater for an external executor still being connected, but slow to respond, etc.?

Reply:
Maybe, if we can get them from the HTTP/2 layer (eg - ping frequency?). But I don't know how valuable they'd be - I think the main decision for a node to make is "this executor is _dead_", and kill the connection. Hopefully that bubbles up from missed pings, and if not we'll do it with an activity timer. There's no node behavioural change before that, it should still dispatch to this executor at max rate. So these stats would be purely for external diagnosis, and may be found more easily by an operator contacting the executor directly.